### PR TITLE
Feature/core types

### PR DESCRIPTION
**Pull request by piiertho**
This implements core types.
This fixes a memory leak when passing GC handled types as icall param.
This introduce a bridge manager to handle bridges glue code in one place.

Edit from @chippmann: Corresponding Entry-Gen PR: https://github.com/utopia-rise/godot-kotlin-entry-generator/pull/7

### CODE REVIEW

**Comment by piiertho** [comment]
This needs JVM_UBYTE registration before merge

File: harness/tests/src/main/kotlin/godot/tests/Invocation.kt [hunk 1]
```diff
@@ -139,8 +189,124 @@ class Invocation : Spatial() {
 
     override fun _onDestroy() {
         invocation.free()
+        navMeshes.clear()
+        navMeshesDictionary.clear()
+        anyToAnyDictionary.clear()
+        nullableArray.clear()
+        nullableDictionary.clear()
+        testArrayAny.clear()
```
Comment (chippmann) [inline harness/tests/src/main/kotlin/godot/tests/Invocation.kt:197]: Why are these needed? Shouldn't they be cleared by the GC?

File: harness/tests/test/unit/test_node_path.gd [hunk 2]
```diff
@@ -0,0 +1,12 @@
+extends "res://addons/gut/test.gd"
+
+
+func test_should_find_camera_node():
+	var invocation_scene = load("res://Spatial.tscn").instance()
+	assert_true(invocation_scene.has_camera_node(), "Should find Camera Node using not path")
```
Comment (chippmann) [inline harness/tests/test/unit/test_node_path.gd:6]: Typo: ... using **node** path

File: kt/godot-library/src/main/kotlin/godot/tests.kt [hunk 3]
```diff
@@ -337,20 +335,30 @@ open class Node : Object() {
...
-        TransferContext.readReturnValue(VariantType.NIL)
     }
 
     open fun removeChild(node: Node) {
         TransferContext.writeArguments(VariantType.OBJECT to node)
         TransferContext.callMethod(rawPtr, NODE_REMOVE_CHILD, VariantType.NIL)
-        TransferContext.readReturnValue(VariantType.NIL)
+    }
+
+    open fun getNode(nodePath: NodePath): Node? {
```
Comment (chippmann) [inline kt/godot-library/src/main/kotlin/godot/tests.kt:352]: I think we should also provide the same method with a string as a param and internally wrap that string into a nodePath for convenience

File: kt/godot-library/src/main/kotlin/godot/tests.kt [hunk 4]
```diff
@@ -337,20 +335,30 @@ open class Node : Object() {
...
+        TransferContext.writeArguments(VariantType.NODE_PATH to nodePath)
+        TransferContext.callMethod(rawPtr, NODE_GET_NODE, VariantType.OBJECT)
+        return TransferContext.readReturnValue(VariantType.OBJECT, true) as Node?
+    }
+
+    open fun getNodeOrNull(nodePath: NodePath): Node? {
```
Comment (chippmann) [inline kt/godot-library/src/main/kotlin/godot/tests.kt:358]: same here

File: kt/godot-library/src/main/kotlin/godot/tests.kt [hunk 5]
```diff
@@ -337,20 +335,30 @@ open class Node : Object() {
...
     open fun removeChild(node: Node) {
         TransferContext.writeArguments(VariantType.OBJECT to node)
         TransferContext.callMethod(rawPtr, NODE_REMOVE_CHILD, VariantType.NIL)
-        TransferContext.readReturnValue(VariantType.NIL)
+    }
+
+    open fun getNode(nodePath: NodePath): Node? {
+        TransferContext.writeArguments(VariantType.NODE_PATH to nodePath)
+        TransferContext.callMethod(rawPtr, NODE_GET_NODE, VariantType.OBJECT)
+        return TransferContext.readReturnValue(VariantType.OBJECT, true) as Node?
```
Comment (chippmann) [inline kt/godot-library/src/main/kotlin/godot/tests.kt:355]: Why is the return type nullable here? Isn't `getNodeOrNull` for nullables and this one should crash if it's null?

File: kt/godot-runtime/src/main/kotlin/godot/core/Dictionary.kt [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    val valVariantType = variantMapper[V::class]
+    checkNotNull(valVariantType) {
+        "Can't create a Dictionary with generic value ${V::class}."
+    }
+    return Dictionary<K, V>(keyVariantType, valVariantType)
+}
+
+inline fun <reified K, reified V> dictionaryOf(vararg args: Pair<K, V>) = Dictionary<K, V>().also {
+    it.putAll(args)
+}
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/Dictionary.kt:335]: Shouldn't we also have convenience functions to convert a Map into a dictionary other than the corresponding constructor?
Like extension functions for Map itself?

File: kt/godot-runtime/src/main/kotlin/godot/core/NativeCoreTypeWeakReference.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package godot.core
+
+import java.lang.ref.WeakReference
+
+internal class NativeCoreTypeWeakReference(referent: NativeCoreType)
+    : WeakReference<NativeCoreType>(referent) {
+    val coreVariantType = referent.coreVariantType
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/NativeCoreTypeWeakReference.kt:7]: doesn't this create a normal reference to `coreVariantType` and thus the ref to `referent` isn't weak anymore?
Souldn't it be (pseudo code): 
```kotlin
val coreVariantType: VariantType
    get() = referent.coreVariantType
```

File: kt/godot-runtime/src/main/kotlin/godot/core/Properties.kt [hunk 8]
```diff
@@ -62,10 +68,27 @@ class KtEnumProperty<T : KtObject, P : Any>(
...
-        require(argsSize == 1) { "Setter should be called with only one argument." }
-        val arg = TransferContext.readSingleArgument(VariantType.JVM_INT)
-        TransferContext.buffer.rewind()
-        kProperty.set(instance, setValueConverter(arg as Int))
+        val arg = extractSetterArgument<Int>()
+        kProperty.set(instance, setValueConverter(arg))
+    }
+}
+
+class KtArrayProperty<T: KtObject, P: Any?>(
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/Properties.kt:76]: This is not used atm. is it needed?

File: kt/godot-runtime/src/main/kotlin/godot/core/Variant.kt [hunk 9]
```diff
@@ -1,9 +1,42 @@
 package godot.core
 
+import godot.core.VariantType.*
 import godot.util.toRealT
 import java.nio.ByteBuffer
 
 
+val variantMapper = mutableMapOf(
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/Variant.kt:8]: Isn't the new `JVM_BYTE` missing?

File: kt/godot-runtime/src/main/kotlin/godot/core/VariantArray.kt [hunk 10]
```diff
@@ -1,57 +1,472 @@
...
+            variantType
+    )
+}
+
+//HELPER
+inline fun <reified T> variantArrayOf(vararg args: T) = VariantArray<T>().also { it.addAll(args) }
+
+/**
+ * Convert an iterable into a GodotArray
+ * Warning: Might be slow if the iterable contains a lot of items because GDNative can only append items one by one
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/VariantArray.kt:466]: Maybe add a Todo here as this is about to change (not sure if only 4.0 or if it will be backported to 3.2

File: kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+package godot.core
+
+import godot.util.IndexedIterator
+import godot.util.RealT
+import godot.util.VoidPtr
+import godot.util.realTVariantType
+
+@Suppress("MemberVisibilityCanBePrivate", "unused")
+class PoolRealArray : NativeCoreType, Iterable<RealT> {
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt:11]: Why are the pool arrays not inheriting mutable collection? just because of the function naming for append and so on? or does it have another reason as well?

File: kt/godot-runtime/src/main/kotlin/godot/runtime/Registration.kt [hunk 12]
```diff
@@ -78,6 +78,32 @@ class ClassBuilderDsl<T : KtObject>(
         )
     }
 
+    fun <P: Any> arrayProperty(
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/runtime/Registration.kt:81]: Not used at the moment. Is this needed? See also my comment on `KtArrayProperty`

**Review by chippmann — CHANGES_REQUESTED** [review]
Awesome work! Some minor comments though

File: kt/godot-library/src/main/kotlin/godot/tests.kt (hunk 3 above)
Comment (piiertho) [reply kt/godot-library/src/main/kotlin/godot/tests.kt:352]: This will be generated with json api, maybe we should do extention function after api generation.

File: kt/godot-library/src/main/kotlin/godot/tests.kt (hunk 5 above)
Comment (piiertho) [reply kt/godot-library/src/main/kotlin/godot/tests.kt:355]: Nop, `get_node` will call `get_node_or_null` and add an addtional error print.

File: kt/godot-runtime/src/main/kotlin/godot/core/Dictionary.kt (hunk 6 above)
Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/core/Dictionary.kt:335]: yep we can add some I think.

File: harness/tests/src/main/kotlin/godot/tests/Invocation.kt (hunk 1 above)
Comment (piiertho) [reply harness/tests/src/main/kotlin/godot/tests/Invocation.kt:197]: Forgot to remove them, was during debugging

File: kt/godot-runtime/src/main/kotlin/godot/core/NativeCoreTypeWeakReference.kt (hunk 7 above)
Comment (CedNaru) [reply kt/godot-runtime/src/main/kotlin/godot/core/NativeCoreTypeWeakReference.kt:7]: It's a normal reference to coreVariantType yes, but this has no effect on referent.

File: kt/godot-runtime/src/main/kotlin/godot/core/Variant.kt (hunk 9 above)
Comment (CedNaru) [reply kt/godot-runtime/src/main/kotlin/godot/core/Variant.kt:8]: You are right, we forgot to add it to the allowed types.

File: kt/godot-runtime/src/main/kotlin/godot/core/VariantArray.kt (hunk 10 above)
Comment (CedNaru) [reply kt/godot-runtime/src/main/kotlin/godot/core/VariantArray.kt:466]: Actually, it can be changed right now.
GDNative didn't have this function, but now we are in a module so I can get the true equal operator from the C++

File: kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt (hunk 11 above)
Comment (CedNaru) [reply kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt:11]: No particular reason, I just kinda implemented Itererable and didn't think about using a more specialized interface.

File: kt/godot-runtime/src/main/kotlin/godot/core/VariantArray.kt (hunk 10 above)
Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/core/VariantArray.kt:466]: We will backport to 3.2 I think, 4.0 should not be so much changes.

File: kt/godot-runtime/src/main/kotlin/godot/core/Properties.kt (hunk 8 above)
Comment (CedNaru) [reply kt/godot-runtime/src/main/kotlin/godot/core/Properties.kt:76]: Was that to store the type of the Array? 
If that's the case, it's not necessary anymore, we already have the type information stored in VariantArray

Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/core/Properties.kt]: was to register array, but is deprecated, I removed it.

File: kt/godot-runtime/src/main/kotlin/godot/core/Variant.kt (hunk 9 above)
Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/core/Variant.kt:8]: done

File: kt/godot-runtime/src/main/kotlin/godot/runtime/Registration.kt (hunk 12 above)
Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/runtime/Registration.kt]: removed

File: kt/godot-runtime/src/main/kotlin/godot/core/Dictionary.kt (hunk 6 above)
Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/core/Dictionary.kt:344]: done

File: kt/godot-library/src/main/kotlin/godot/tests.kt (hunk 3 above)
Comment (chippmann) [reply kt/godot-library/src/main/kotlin/godot/tests.kt:352]: Ok for me if we do it there

File: kt/godot-library/src/main/kotlin/godot/tests.kt (hunk 5 above)
Comment (chippmann) [reply kt/godot-library/src/main/kotlin/godot/tests.kt:355]: I see. Thx for the clarification

File: kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt (hunk 11 above)
Comment (chippmann) [reply kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt:11]: shouldn't we then not use mutable collection?
Then one can use all of the kotlin collection function

**Review by chippmann — APPROVED** [review]

File: kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt (hunk 11 above)
Comment (piiertho) [reply kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt:11]: Maybe we can do it after merging this as an enhancement ? This would allow us to split last tasks for alpha.

Comment (chippmann) [reply kt/godot-runtime/src/main/kotlin/godot/core/pool/PoolRealArray.kt:11]: Ok for me

**Review by CedNaru — APPROVED** [review]

**Review by raniejade — APPROVED** [review]
Sorry for the delay!
